stl/inc/semaphore

```diff
@@ -0,0 +1,85 @@
+// semaphore standard header
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+
+#ifndef _SEMAPHORE_
+#define _SEMAPHORE_
+#include <yvals_core.h>
+#if _STL_COMPILER_PREPROCESSOR
+
+#if !_HAS_CXX20
+#pragma message("The contents of <semaphore> are available only with C++20 or later.")
+#else // ^^^ !_HAS_CXX20 / _HAS_CXX20 vvv
+#include <atomic>
+#include <cassert>
+#include <chrono>
+#include <cstddef>
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#undef new
+
+_STD_BEGIN
+template <ptrdiff_t _LeastMaxValue = 256>
+class counting_semaphore {
+public:
+    static constexpr ptrdiff_t max() noexcept {
+        // _LeastMaxValue is used here so that things such as binary_semaphore are
+        return _LeastMaxValue;
+    }
+    constexpr explicit counting_semaphore(ptrdiff_t _Desired) {
+        _STL_ASSERT(_Desired >= 0 && _Desired <= max(), "Desired value for the semaphore falls outside of semaphore range");
+        atomic_init(&_Counter, _Desired);
+    }
+    ~counting_semaphore()                         = default;
+    counting_semaphore(const counting_semaphore&) = delete;
+    counting_semaphore& operator=(const counting_semaphore&) = delete;
+
+    void release(ptrdiff_t _Update = 1) {
+        _STL_ASSERT(_Update >= 0 && _Update <= (max() - _Counter.load()), "The semaphore has been released more times than acquired");
+        _Counter.fetch_add(_Update, memory_order_release);
+    }
+    void acquire() {
+        while (!try_acquire()) {
+        }
+    }
+    bool try_acquire() noexcept {
+        ptrdiff_t _Old     = _Counter.load(memory_order_acquire);
+        ptrdiff_t _Desired = _Old - 1;
+        if (_Old > 0 && _Counter.compare_exchange_strong(_Old, _Desired, memory_order_acquire)) {

```

Comment:
`memory_order_acquire` is certainly incorrect; this is publishing the reduced counter value to other threads. If you want to use weaker than sequentially consistent atomics you need to do the writeup proving that the result is correct according to the memory model.

Reply:
`memory_order_acquire` is fine here, since only `release` is said to _strongly happen before_ `acquire`, there's no such relationship between different `acquires`.

Reply:
This piece is more tricky for `barrier`, when `arrive`/`arrive_and_wait`, where publishing reduced value indeed matters, since one of them will call completion function...

Reply:
For `latch` too, by the way, `count_down` is release decrement, but `arrive_and_wait` is acq_rel if the value of decrement is also used to unblock immediately.

Reply:
Yhea, I was mostly confused at why memory_order_acquire here was said to be wrong, esp. since the semantics of semaphore and the terminology used explicitly make clear that acquire is fine since you're performing, well, an "acquire" operation. BTW, @AlexGuteniev your PR for semaphore is remarkably similar to what my "final" design that I never actually made a PR for, for semaphore, looks like, and seems to fix all of the problems I had making it (as well as gets rid of the main problem I encounted, which was the WaitOnAddress stuff). Thanks for putting in the effort that I was too lazy to go through

Reply:
Ok, this one is more complex than I anticipated originally.

See GH-1133 as the main problem, that is not resolved yet.

And also see https://stackoverflow.com/q/63215461/2945027 as another complication I encountered (but I think it is resolved in comments there).